src/couch_views/src/couch_views_batch.erl

```diff
@@ -0,0 +1,80 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+% http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_views_batch).
+
+
+-export([
+    start/1,
+    success/3,
+    failure/1
+]).
+
+
+-include_lib("couch_mrview/include/couch_mrview.hrl").
+
+
+-callback start(
+            Mrst::#mrst{},
+            State::term()
+        ) -> {NewState::term(), BatchSize::pos_integer()}.
+
+-callback success(
+            Mrst::#mrst{},
+            TxSize::non_neg_integer(),
+            DocsRead::non_neg_integer(),
```

Comment:
The `couch_rate:success/2` callback used `WrittenDocs` argument. Would it be possible to pass `DocsWrite` to the callback?

Reply:
If you look at the old implementation `WrittenDocs` is exactly the same value as the new `DocsRead`, except that `DocsRead` is taken directly from `length(DocAcc)` where as `WrittenDocs` is calculated while iterating over `DocAcc`.

Reply:
My understanding is that map function can emit 0 or more documents, therefore `DocsRead /= DocsWritten`:

- DocsRead  - `length(DocAcc)`
- DocsWritten - `length(MappedDocs)`

In `couch_rate` the `WrittenDocs` was a result of iterating over `MappedDocs` https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_views/src/couch_views_indexer.erl#L205

The reason I calculated it during iteration is to avoid extra `O(n)` cost on `length(List)`.

Reply:
A map function can emit 0 or more rows for each of the defined views. For every doc we have a result set regardless of whether it has rows or not (because we have to remember we may be removing a document from a view). Regardless of whether a doc emits rows or not, it still has an entry in `MappedDocs` and `length(DocAcc) == length(MappedDocs)`.

I could see counting up the number of rows emitted for each document as maybe being an interesting value to have access to.

Reply:
I've tweaked the API to include the total number of emitted KVs and switched to using a map with a defined type exported from `couch_views_batch`. Let me know if you can think of any other useful values or if those three are good enough for now.